Let's use what we've learned to play a mage duel!

Imagine your wizard has 6d4 health points and you have spells that do 6d4 damage. "6d4" means rolling six 4-sided dice and summing the result.

Your opposing mage has 4d6 health points and spells that do 4d6. "4d6" means rolling four six-sided dice and summing the result.

Exercises

Simulate mage duels to answer who is the more powerful mage?

1. Before running simulations, do you have a hypothesis of which mage will win? 
- Do you have a hunch? Write it down. This is your first exercise.
2. Simulate 10 mage duels. Is there a clear winner? 
- Run that 10 duel simulation again. Was the answer similar?
3. Do the results change much at 100 duels?
4. Now, simulate 10,000 mage duels. Is there a clear winner?

I believe that our wizard will be more powerful, because the minimum dice total value of both hp and attack (6) is higher than the enemy minimum hp and attack value (4).

In [174]:
#expected values of dice rolls
ev_6d4 = (1*(.25) + 2*(.25) + 3*(.25) + 4*(.25)) * 6
ev_4d6 = (1*(1/6) + 2*(1/6) + 3*(1/6) + 4*(1/6) + 5*(1/6) + 6*(1/6)) * 4
print(f'expected value of 6d4 is: {ev_6d4}')
print(f'expected value of 4d6 is: {ev_4d6}')

expected value of 6d4 is: 15.0
expected value of 4d6 is: 14.0


In [1]:
import numpy as np
import pandas as pd

In [183]:
n_simulations = n_rows = 10
# create an empty list to store our wins list
wins = []
#wizard_hp is 6d4
n_wizard_dice = n_cols_w = 6
# get the wizard starting hp for each simulation
wizard_starting_hp = np.random.choice([1,2,3,4], n_cols_w * n_rows).reshape(
    n_rows, n_cols_w).sum(axis=1)
#enemy_hp is 4d6
n_enemy_dice = n_cols_e = 4
# get the enemy starting hp for each simulation
enemy_starting_hp = np.random.choice([1,2,3,4,5,6], n_cols_e * n_rows).reshape(
    n_rows, n_cols_e).sum(axis=1)
# print(wizard_starting_hp)

for i in range(n_simulations):
#     print(f'your wizard\'s starting hp is {wizard_starting_hp[i]}')
#     print(f'the enemy wizard\'s starting hp is: {enemy_starting_hp[i]}')
    wizard_hp = wizard_starting_hp[i]
    enemy_hp = enemy_starting_hp[i]
    while True:
        # loop through attacks until one hp is 0
        wizard_attack = np.random.choice([1,2,3,4], n_cols_w).sum()
        enemy_attack  = np.random.choice([1,2,3,4,5,6], n_cols_e).sum()
        wizard_hp -= enemy_attack
#         print(f'enemy attacks for: {enemy_attack} wizard hp is: {wizard_hp}')
        enemy_hp -= wizard_attack
#         print(f'wizard attacks for: {wizard_attack} enemy hp is: {enemy_hp}')
        # if one hp is 0 then break loop, store result, and move to next simulation
        if (wizard_hp <= 0) and (enemy_hp <= 0):
            wins.append('t') 
#             print('tie, nobody wins')
            break
        elif (wizard_hp <= 0):
            wins.append('e') 
#             print('enemy wins')
            break
        elif (enemy_hp <= 0):
            wins.append('w')
#             print('your wizard wins')
            break

# print(wins)
# percentages
wizard_wins = round(wins.count('w') / len(wins), 4)
print(f'our wizard win percentage is: {wizard_wins}')
enemy_wins = round(wins.count('e') / len(wins), 4)
print(f'enemy win percentage is: {enemy_wins}')
ties = round(wins.count('t') / len(wins), 4)
print(f'tie percentage is: {ties}')

our wizard win percentage is: 0.3
enemy win percentage is: 0.1
tie percentage is: 0.6


after running 10 simulations most common outcome is tie, after that it appears our wizard wins more often

In [184]:
# 100 duels
n_simulations = n_rows = 100
# create an empty list to store our wins list
wins = []
#wizard_hp is 6d4
n_wizard_dice = n_cols_w = 6
# get the wizard starting hp for each simulation
wizard_starting_hp = np.random.choice([1,2,3,4], n_cols_w * n_rows).reshape(
    n_rows, n_cols_w).sum(axis=1)
#enemy_hp is 4d6
n_enemy_dice = n_cols_e = 4
# get the enemy starting hp for each simulation
enemy_starting_hp = np.random.choice([1,2,3,4,5,6], n_cols_e * n_rows).reshape(
    n_rows, n_cols_e).sum(axis=1)
# print(wizard_starting_hp)

for i in range(n_simulations):
#     print(f'your wizard\'s starting hp is {wizard_starting_hp[i]}')
#     print(f'the enemy wizard\'s starting hp is: {enemy_starting_hp[i]}')
    wizard_hp = wizard_starting_hp[i]
    enemy_hp = enemy_starting_hp[i]
    while True:
        # loop through attacks until one hp is 0
        wizard_attack = np.random.choice([1,2,3,4], n_cols_w).sum()
        enemy_attack  = np.random.choice([1,2,3,4,5,6], n_cols_e).sum()
        wizard_hp -= enemy_attack
#         print(f'enemy attacks for: {enemy_attack} wizard hp is: {wizard_hp}')
        enemy_hp -= wizard_attack
#         print(f'wizard attacks for: {wizard_attack} enemy hp is: {enemy_hp}')
        # if one hp is 0 then break loop, store result, and move to next simulation
        if (wizard_hp <= 0) and (enemy_hp <= 0):
            wins.append('t') 
#             print('tie, nobody wins')
            break
        elif (wizard_hp <= 0):
            wins.append('e') 
#             print('enemy wins')
            break
        elif (enemy_hp <= 0):
            wins.append('w')
#             print('your wizard wins')
            break

# print(wins)
# percentages
wizard_wins = round(wins.count('w') / len(wins), 4)
print(f'our wizard win percentage is: {wizard_wins}')
enemy_wins = round(wins.count('e') / len(wins), 4)
print(f'enemy win percentage is: {enemy_wins}')
ties = round(wins.count('t') / len(wins), 4)
print(f'tie percentage is: {ties}')

our wizard win percentage is: 0.34
enemy win percentage is: 0.1
tie percentage is: 0.56


after 100 duels: tie is the most frequent outcome, after that our wizard wins most often

In [190]:
# 10_000 duels
n_simulations = n_rows = 10_000
# create an empty list to store our wins list
wins = []
#wizard_hp is 6d4
n_wizard_dice = n_cols_w = 6
# get the wizard starting hp for each simulation
wizard_starting_hp = np.random.choice([1,2,3,4], n_cols_w * n_rows).reshape(
    n_rows, n_cols_w).sum(axis=1)
#enemy_hp is 4d6
n_enemy_dice = n_cols_e = 4
# get the enemy starting hp for each simulation
enemy_starting_hp = np.random.choice([1,2,3,4,5,6], n_cols_e * n_rows).reshape(
    n_rows, n_cols_e).sum(axis=1)
# print(wizard_starting_hp)

for i in range(n_simulations):
#     print(f'your wizard\'s starting hp is {wizard_starting_hp[i]}')
#     print(f'the enemy wizard\'s starting hp is: {enemy_starting_hp[i]}')
    wizard_hp = wizard_starting_hp[i]
    enemy_hp = enemy_starting_hp[i]
    while True:
        # loop through attacks until one hp is 0
        wizard_attack = np.random.choice([1,2,3,4], n_cols_w).sum()
        enemy_attack  = np.random.choice([1,2,3,4,5,6], n_cols_e).sum()
        wizard_hp -= enemy_attack
#         print(f'enemy attacks for: {enemy_attack} wizard hp is: {wizard_hp}')
        enemy_hp -= wizard_attack
#         print(f'wizard attacks for: {wizard_attack} enemy hp is: {enemy_hp}')
        # if one hp is 0 then break loop, store result, and move to next simulation
        if (wizard_hp <= 0) and (enemy_hp <= 0):
            wins.append('t') 
#             print('tie, nobody wins')
            break
        elif (wizard_hp <= 0):
            wins.append('e') 
#             print('enemy wins')
            break
        elif (enemy_hp <= 0):
            wins.append('w')
#             print('your wizard wins')
            break

# print(wins)
# percentages
wizard_wins = round(wins.count('w') / len(wins), 4)
print(f'our wizard win percentage is: {wizard_wins}')
enemy_wins = round(wins.count('e') / len(wins), 4)
print(f'enemy win percentage is: {enemy_wins}')
ties = round(wins.count('t') / len(wins), 4)
print(f'tie percentage is: {ties}')

our wizard win percentage is: 0.3452
enemy win percentage is: 0.1727
tie percentage is: 0.4821


Our wizard is the clear winner with around 35% win rate, tie is around 48% and the enemy wins around 17%. This is the outcome that I was expecting.